src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs

```diff
@@ -155,22 +151,34 @@ private static bool HasPath(Location location)
             return !string.IsNullOrEmpty(location.GetLineSpan().Path);
         }
 
+        private static readonly Uri _fileRoot = new Uri("file:///");
+
         private static string GetUri(string path)
         {
             Debug.Assert(!string.IsNullOrEmpty(path));
 
+            // Note that in general, these "paths" are opaque strings to be 
+            // interpreted by resolvers (see SyntaxTree.FilePath documentation).
+            
+            // Common case: absolute path -> absolute URI
             Uri uri;
+            if (Uri.TryCreate(path, UriKind.Absolute, out uri))
+            {
+                // Note that URI.ToString() does not, for example, escape spaces to %20.
+                return uri.AbsoluteUri;
+            }
 
-            if (!Uri.TryCreate(path, UriKind.RelativeOrAbsolute, out uri))
+            // First fallback attempt: interpret as relative reference if possible
+            // (perhaps the resolver works that way).
+            if (Uri.TryCreate(path, UriKind.Relative, out uri))
             {
-                // The only constraint on paths are that they can be interpreted by
-                // various resolvers so there is no guarantee we can turn the arbitrary string
-                // in to a URI. If our attempt to do so fails, use the original string as the
-                // "URI".
-                return path;
+                // There is no AbsoluteUri equivalent for relative URI references and ToString() 
+                // won't escape without this relative -> absolute -> relative trick.
+                return _fileRoot.MakeRelativeUri(new Uri(_fileRoot, uri)).ToString();
             }
 
-            return uri.ToString();
+            // Last resort: UrlEncode the whole opaque string as a relative reference with one part.
+            return System.Net.WebUtility.UrlEncode(path);
```

Comment:
ok.

---
In reply to: [63799241](https://github.com/dotnet/roslyn/pull/11390#discussion_r63799241) [](ancestors = 63799241)